spec/builtins.test.sh

```diff
@@ -1,160 +1,188 @@
+# targ=$TMP/cd-symtarget
+# lnk=$TMP/cd-symlink
+# mkdir -p $targ
+# ln -s $targ $lnk
+# cd -P $targ
+# test $PWD = "$TMP/cd-symtarget" && echo OK
+# cd $TMP
+# rmdir $targ
+# rm $lnk
+# # stdout: OK
+# 
+# ### cd to symlink default behavior
+# targ=$TMP/cd-symtarget
+# lnk=$TMP/cd-symlink
+# mkdir -p $targ
+# ln -s $targ $lnk
+# cd $lnk
+# test $PWD = "$TMP/cd-symlink" && echo OK
+# cd $TMP
+# rmdir $targ
+# rm $lnk
+# # stdout: OK
+# 
+# ### cd to symlink with -L
+# targ=$TMP/cd-symtarget
+# lnk=$TMP/cd-symlink
+# mkdir -p $targ
+# ln -s $targ $lnk
+# cd -L $lnk
+# test $PWD = "$TMP/cd-symlink" && echo OK
+# cd $TMP
+# rmdir $targ
+# rm $lnk
+# # stdout: OK
+# 
+# ### cd to symlink with -P
+# targ=$TMP/cd-symtarget
+# lnk=$TMP/cd-symlink
+# mkdir -p $targ
+# ln -s $targ $lnk
+# cd -P $lnk
+# test $PWD = "$TMP/cd-symtarget" && echo OK
+# cd $TMP
+# rmdir $targ
+# rm $lnk
+# # stdout: OK
+# 
+# ### pushd/popd
+# set -o errexit
+# cd /
+# pushd $TMP
+# popd
+# pwd
+# # status: 0
+# # N-I dash/mksh status: 127
+# 
+# ### Eval
+# eval "a=3"
+# echo $a
+# # stdout: 3
+# 
+# ### Source
+# lib=$TMP/spec-test-lib.sh
+# echo 'LIBVAR=libvar' > $lib
+# . $lib  # dash doesn't have source
+# echo $LIBVAR
+# # stdout: libvar
+# 
+# ### Source nonexistent
+# source /nonexistent/path
+# echo status=$?
+# # stdout: status=1
+# # OK dash stdout: status=127
+# 
+# ### Source with no arguments
+# source
+# echo status=$?
+# # stdout: status=1
+# # OK bash stdout: status=2
+# # OK dash stdout: status=127
+# 
+# ### Source with extra arguments
+# lib=$TMP/to-be-sourced
+# echo 'echo $@' > $lib
+# . $lib foo bar  # dash doesn't have source
+# # stdout: foo bar
+# # N-I dash stdout:
 
-### exec builtin with redirects
-exec 1>&2
-echo 'to stderr'
-# stdout-json: ""
-# stderr: to stderr
-
-### exec builtin with here doc
-# This has in a separate file because both code and data can be read from
-# stdin.
-$SH spec/builtins-exec-here-doc-helper.sh
-# stdout-json: "x=one\ny=two\nDONE\n"
-
-### cd and $PWD
-cd /
-echo $PWD
-# stdout: /
-
-### $OLDPWD
-cd /
-cd $TMP
-echo "old: $OLDPWD"
-cd -
-# stdout-json: "old: /\n/\n"
-
-### pwd
-cd /
-pwd
-# stdout: /
-
-### pwd -P
-mkdir -p $TMP/symtarg
-ln -s $TMP/symtarg $TMP/symlink
-cd $TMP/symlink
-basename $(pwd -P)
-cd $TMP
-rmdir $TMP/symtarg
-rm $TMP/symlink
-# stdout: symtarg
-
-### cd with no arguments
-HOME=$TMP/home
-mkdir -p $HOME
-cd
-test $(pwd) = "$HOME" && echo OK
-# stdout: OK
-
-### cd to nonexistent dir
-cd /nonexistent/dir
-echo status=$?
-# stdout: status=1
-# OK dash/mksh stdout: status=2
-
-### cd away from dir that was deleted
-dir=$TMP/cd-nonexistent
-mkdir -p $dir
-cd $dir
-rmdir $dir
-cd $TMP
-echo $(basename $OLDPWD)
-echo status=$?
+### Source from inside function
+lib=$TMP/to-be-sourced
+echo 'echo $@' > $lib
+echo 'local foo=foo_val' >> $lib
+f() { . $lib; . $lib args to src; echo $foo; }
+f args to func
 ## STDOUT:
-cd-nonexistent
-status=0
+args to func
+args to src
+foo_val
 ## END
+## N-I dash
```

Comment:
@andychu `bash` and `mksh` pass this test as-is; note that `args to func` is printed even though the script is sourced without `$@`. (Side-note: to get the "N-I" status for Dash, do I need to add a `stdout-json` to the `N-I` line, or can I specify a multi-line `STDOUT` for a `N-I` entry?)

So sourcing, at least, *does* need to create some kind of "optional" `argv` entry. I still think the `__enter__`/`__exit__` pattern would be the best way to handle scope push/pop logic; would you be amenable to that change?